libbeat/common/mapstr.go

```diff
@@ -257,6 +257,37 @@ func (m MapStr) Format(f fmt.State, c rune) {
 	io.WriteString(f, debugM.String())
 }
 
+// Expand does the reverse of Flatten - expands dotted text and return an expanded MapStr.
+//
+// Example:
+//   "hello.world": "test"
+//
+// This is converted to:
+//   "hello": MapStr{"world": "test" }
+//
+func (m MapStr) Expand() (MapStr, error) {
+	e := MapStr{}
+	f := m.Flatten()
+	keys := make([]string, 0, len(m))
+	for k := range f {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for i := len(keys) - 1; i >= 0; i-- {
+		k := keys[i]
+		old, err := e.Put(k, f[k])
+		if err != nil {
+			return e, err
+		}
+		_, oldIsMap := tryToMapStr(old)
+		_, newIsMap := tryToMapStr(f[k])
+		if oldIsMap != newIsMap {
+			return e, fmt.Errorf("incompatible types expanded for %q: old: %T new: %T", k, old, f[k])
+		}
+	}
+	return e, nil
+}
+
```

Comment:
When using the `Expand()` method to expand logging documents returning an error for the method would lead to loosing logs. Should we instead drop one of the fields causing the error (since the fields are ordered it would be deterministic) and maybe log a warning?

Reply:
If an error in the pipeline processing happens, I would expect that Filebeat would still index the event with the raw and invalid JSON as a string in the `message` field and with the `error.message` and `error.type` fields set.

See also this demo:

```
$ ./filebeat -c /dev/null --strict.perms=false -E output.console.enabled=true -E filebeat.inputs='[{type: stdin, enabled:true, json.keys_under_root:true, json.add_error_key:true}]'
foo
{"@timestamp":"2020-11-06T19:54:34.117Z","@metadata":{"beat":"filebeat","type":"_doc","version":"7.9.3"},"agent":{"ephemeral_id":"6bea7e71-64f3-4f05-bad9-561b33f93159","id":"d68943aa-0741-48d0-9c73-940d2cb14035","name":"ix.lan","type":"filebeat","version":"7.9.3","hostname":"ix.lan"},"message":"foo","log":{"offset":0,"file":{"path":""}},"error":{"message":"Error decoding JSON: invalid character 'o' in literal false (expecting 'a')","type":"json"},"input":{"type":"stdin"},"ecs":{"version":"1.5.0"},"host":{"name":"ix.lan"}}
```

Reply:
@simitt good point! I hadn't thought that through before: although libbeat will log a warning and continue indexing with the original document, Elasticsearch will internally expand the JSON object fields for field mapping, and this would lead to the same kind of conflict/error that we encounter in the processor, leading to the document not being indexed.

If we still drop conflicting fields in the error case, I wonder if it would be obvious enough from the document that there was an issue in processing.

@felixbarny At this point we've already decoded the JSON, and it's technically valid. I think it would be a bit surprising to then stringify the JSON and put it all back in the `message` field.

Unless... instead of adding a new processor, we add an option to the `json` input type instead? i.e. instead of

`./filebeat -e -c /dev/null --strict.perms=false -E output.console.enabled=true -E filebeat.inputs='[{type: stdin, enabled:true, json.keys_under_root:true}]' -E processors='[{dot_expand:{}}]' -d '*'`

we could have

`./filebeat -e -c /dev/null --strict.perms=false -E output.console.enabled=true -E filebeat.inputs='[{type: stdin, enabled:true, json.keys_under_root:true, json.expand_fields:true}]' -d '*'`

Then an error to expand fields would be be an error to decode, and the behaviour you describe would be natural.

Reply:
Having it directly in `json.expand_fields` would be nice. But it would then also have to be a property in the [`decode_json_fields`](https://www.elastic.co/guide/en/beats/filebeat/current/decode-json-fields.html) processor.